iosslsniffer/sniffer.py

```diff
@@ -0,0 +1,90 @@
+import logging
+import os
+import posixpath
+from dataclasses import dataclass
+from typing import Tuple
+
+import coloredlogs
+from pymobiledevice3.lockdown import LockdownClient
+from pymobiledevice3.services.syslog import SyslogService
+
+from iosslsniffer.exceptions import ParseFailedError
+from iosslsniffer.ssl_session import SSLSession
+from iosslsniffer.syslog_parser import SyslogParser
+
+coloredlogs.install(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Filters:
+    pids: Tuple = None
+    process_names: Tuple = None
+    images: Tuple = None
+
+
+@dataclass
+class EntryHash:
+    pid: int
+    process_name: str
+    image: str
+    domain: str
+
+
+class Sniffer:
+    def __init__(self, lockdown: LockdownClient, out_file: str = 'tmp.pcapng', filters=None):
+        self._filters = filters
+        self._out_file = out_file
+        self.lockdown = lockdown
+        self.os_trace_service = SyslogService(self.lockdown)
+        self.syslog_parser = SyslogParser()
+        self.active_fd = {}
+
+    def sniff(self) -> None:
+        for entry in self.os_trace_service.watch():
+            if 'CFNetwork' not in entry:
+                continue
+
+            if 'done using Connection' in entry:
+                fd = int(entry.split()[-1])
+                if fd in self.active_fd:
+                    entry_hash, stream = self.active_fd[fd]
+                    logger.info(
+                        f'Saving recording for {entry_hash.process_name}({entry_hash.pid})  💾  {entry_hash.domain}')
+                    stream.dump(self._out_file)
+                    del self.active_fd[fd]
+
+            if 'SSL' not in entry:
+                continue
+            try:
+                parsed = self.syslog_parser.parse(entry)
+            except ParseFailedError:
+                continue
+
+            entry_hash = EntryHash(parsed.pid,
+                                   posixpath.basename(parsed.filename),
+                                   os.path.basename(parsed.image_name),
+                                   parsed.domain)
+
+            if not self.should_keep(entry_hash):
+                continue
+
+            fd = parsed.fd
+            if fd not in self.active_fd:
+                logger.info(
+                    f'Recording SSL traffic {entry_hash.process_name}({parsed.pid}) ➡️  {entry_hash.domain}')
+                self.active_fd[fd] = (entry_hash, SSLSession(fd, parsed.local, parsed.peer))
+
+            self.active_fd[fd][1].process(parsed)
+
+    def should_keep(self, entry_hash: EntryHash) -> bool:
+        """Filter out entry if one of the criteria specified (pid,image,process_name)"""
+        if self._filters.pids and entry_hash.pid in self._filters.pids:
```

Comment:
isn't this the opposite of what we want? if user included a pid list we should keep it if `entry_hash.pid in self._filters.pids`